modules/network/src/main/java/org/apache/ignite/network/DefaultMessagingService.java

```diff
@@ -0,0 +1,363 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.network;
+
+import static java.util.concurrent.CompletableFuture.failedFuture;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.ignite.internal.network.NetworkMessagesFactory;
+import org.apache.ignite.internal.network.message.InvokeRequest;
+import org.apache.ignite.internal.network.message.InvokeResponse;
+import org.apache.ignite.internal.network.message.ScaleCubeMessage;
+import org.apache.ignite.internal.network.netty.ConnectionManager;
+import org.apache.ignite.internal.network.netty.NettySender;
+import org.apache.ignite.lang.NodeStoppingException;
+import org.jetbrains.annotations.Nullable;
+
+/** Default messaging service implementation. */
+public class DefaultMessagingService extends AbstractMessagingService {
+    /** Network messages factory. */
+    private final NetworkMessagesFactory factory = new NetworkMessagesFactory();
+
+    /** Topology service. */
+    private final TopologyService topologyService;
+
+    /** Connection manager that provides access to {@link NettySender}. */
+    private volatile ConnectionManager connectionManager;
+
+    /**
+     * This node's local socket address. Not volatile because it's piggybacked by the {@link #connectionManager} and connection manager
```

Comment:
I would like to rephrase this a little bit:
```
Not volatile, because access is guarded by volatile reads/writes to the {@code connectionManager} variable.
```